keyboards/keychron/q4/q4.c

```diff
@@ -0,0 +1,76 @@
+/* Copyright 2022 @ Keychron (https://www.keychron.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "q4.h"
+
+const matrix_row_t matrix_mask[] = {
+    0b11111111111111,
+    0b11111111111111,
+    0b11111111111111,
+    0b11111111111111,
+    0b11111111101111,
+};
+
+#ifdef DIP_SWITCH_ENABLE
+
+bool dip_switch_update_kb(uint8_t index, bool active) {
+    if (!dip_switch_update_user(index, active)) { return false;}
+    if (index == 0) {
+        default_layer_set(1UL << (active ? 1 : 0));
+    }
+    return true;
+}
+
+#endif // DIP_SWITCH_ENABLE
+
+#if defined(RGB_MATRIX_ENABLE) && defined(CAPS_LOCK_LED_INDEX)
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+    if (!process_record_user(keycode, record)) { return false; }
+    switch (keycode) {
+#ifdef RGB_MATRIX_ENABLE
+        case RGB_TOG:
+            if (record->event.pressed) {
+                switch (rgb_matrix_get_flags()) {
+                    case LED_FLAG_ALL: {
+                        rgb_matrix_set_flags(LED_FLAG_NONE);
+                        rgb_matrix_set_color_all(0, 0, 0);
+                    } break;
+                    default: {
+                        rgb_matrix_set_flags(LED_FLAG_ALL);
+                    } break;
+                }
+            }
+            if (!rgb_matrix_is_enabled()) {
+                rgb_matrix_set_flags(LED_FLAG_ALL);
+                rgb_matrix_enable();
+            }
+            return false;
+#endif
+    }
+    return true;
+}
+__attribute__((weak)) void rgb_matrix_indicators_advanced_user(uint8_t led_min, uint8_t led_max) {
+    // RGB_MATRIX_INDICATOR_SET_COLOR(index, red, green, blue);
```

Comment:
```suggestion
bool rgb_matrix_indicators_kb(void) {
    if (!rgb_matrix_indicators_user()) {
        return false;
    }
    // RGB_MATRIX_INDICATOR_SET_COLOR(index, red, green, blue);
```

Reply:
Implementing this suggested code spat out this error:
```
Compiling: keyboards/keychron/q4/q4.c                                                              keyboards/keychron/q4/q4.c:66:6: error: conflicting types for 'rgb_matrix_indicators_advanced_kb'
   66 | bool rgb_matrix_indicators_advanced_kb(void) {
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from quantum/quantum.h:36,
                 from keyboards/keychron/q4/q4.h:19,
                 from keyboards/keychron/q4/q4.c:17:
quantum/rgb_matrix/rgb_matrix.h:136:6: note: previous declaration of 'rgb_matrix_indicators_advanced_kb' was here
  136 | bool rgb_matrix_indicators_advanced_kb(uint8_t led_min, uint8_t led_max);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
keyboards/keychron/q4/q4.c: In function 'rgb_matrix_indicators_advanced_kb':
keyboards/keychron/q4/q4.c:67:10: error: too few arguments to function 'rgb_matrix_indicators_advanced_user'
   67 |     if (!rgb_matrix_indicators_advanced_user()) {
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from quantum/quantum.h:36,
                 from keyboards/keychron/q4/q4.h:19,
                 from keyboards/keychron/q4/q4.c:17:
quantum/rgb_matrix/rgb_matrix.h:137:6: note: declared here
  137 | bool rgb_matrix_indicators_advanced_user(uint8_t led_min, uint8_t led_max);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from quantum/quantum.h:36,
                 from keyboards/keychron/q4/q4.h:19,
                 from keyboards/keychron/q4/q4.c:17:
quantum/rgb_matrix/rgb_matrix.h:84:14: error: 'led_min' undeclared (first use in this function)
   84 |     if (i >= led_min && i < led_max) {             \
      |              ^~~~~~~
keyboards/keychron/q4/q4.c:72:9: note: in expansion of macro 'RGB_MATRIX_INDICATOR_SET_COLOR'
   72 |         RGB_MATRIX_INDICATOR_SET_COLOR(CAPS_LOCK_LED_INDEX, 255, 255, 255);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
quantum/rgb_matrix/rgb_matrix.h:84:14: note: each undeclared identifier is reported only once for each function it appears in
   84 |     if (i >= led_min && i < led_max) {             \
      |              ^~~~~~~
keyboards/keychron/q4/q4.c:72:9: note: in expansion of macro 'RGB_MATRIX_INDICATOR_SET_COLOR'
   72 |         RGB_MATRIX_INDICATOR_SET_COLOR(CAPS_LOCK_LED_INDEX, 255, 255, 255);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
quantum/rgb_matrix/rgb_matrix.h:84:29: error: 'led_max' undeclared (first use in this function)
   84 |     if (i >= led_min && i < led_max) {             \
      |                             ^~~~~~~
keyboards/keychron/q4/q4.c:72:9: note: in expansion of macro 'RGB_MATRIX_INDICATOR_SET_COLOR'
   72 |         RGB_MATRIX_INDICATOR_SET_COLOR(CAPS_LOCK_LED_INDEX, 255, 255, 255);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 [ERRORS]
 |
 |
 |
make[1]: *** [builddefs/common_rules.mk:360: .build/obj_keychron_q4_ansi_v2_keychron/keyboards/keychron/q4/q4.o] Error 1
Make finished with errors
```
To rectify this, I had to replace this callback with 
```
bool rgb_matrix_indicators_advanced_kb(uint8_t led_min, uint8_t led_max) {
    if (!rgb_matrix_indicators_advanced_user(led_min, led_max)) {
        return false;
    }
```
which didn't spat out any errors after using this callback.
```
Compiling: keyboards/keychron/q4/q4.c                                                               [OK]
Compiling: quantum/via.c                                                                            [OK]
Linking: .build/keychron_q4_ansi_v2_keychron.elf                                                    [OK]
Creating binary load file for flashing: .build/keychron_q4_ansi_v2_keychron.bin                     [OK]
Creating load file for flashing: .build/keychron_q4_ansi_v2_keychron.hex                            [OK]

Size after:
   text    data     bss     dec     hex filename
      0   48468       0   48468    bd54 keychron_q4_ansi_v2_keychron.bin

Copying keychron_q4_ansi_v2_keychron.bin to qmk_firmware folder                                     [OK]
(Firmware size check does not yet support STM32L432; skipping)
```